src/gc/impl/conservative/gc.d

```diff
@@ -1667,7 +1673,7 @@ struct Gcx
                     recoverNextPage(bin);
                 }
             }
-            else
+            else if (usedSmallPages > 0)
```

Comment:
I think usedSmallPages is not ever going to be 0 here.

Looking at the if statement above:

```D
if (!lowMem && (disabled || usedSmallPages < smallCollectThreshold))
```

Arguably, `lowMem` is going to be false (always is on 64-bit). If it's false, then the condition becomes `disabled || usedSmallPages < smallCollectThreshold`. If not disabled, and `usedSmallPages < smallCollectThreshold`, then 0 would be included in that range.

Really the only way we get here is if the entire memory space is allocated with large pages. Which should NEVER happen, at the very minimum you are going to have some thread-related allocations.

Reply:
This is just meant as a detection whether any pool for small objects is allocated, i.e. it avoids running a collection (and starting the threads for parallel marking) for the very first allocation.

Reply:
Right, but it won't get here. The very first allocation should go into the if branch.

In the first allocation, `lowMem` is false, disabled is false, and usedSmallPages is 0. `if(!false && (false || 0 < smallCollectThreshold))` should be true, so it will never get to this else if.

Or am I reading it wrong?

Reply:
`smallCollectThreshold` is 0, too, see https://github.com/dlang/druntime/blob/3d44d2affda45e638ed52b0c9658435b3418d7f8/src/gc/impl/conservative/gc.d#L1261

Reply:
My goodness. what the hell is it for then?

Reply:
OK, I see it's based on some mechanism to tune the collection frequency based on used/allocated memory. I had thought it was a constant to prevent exactly what you were trying to prevent. So much going on in here!